File: ServidorArgentum.NET/src/clsByteQueue.vb
Option Strict On
Option Explicit On

Friend Class clsByteQueue
    ' Exception codes
    Private Const NOT_ENOUGH_DATA As Integer = vbObjectError + 9
    Private Const NOT_ENOUGH_SPACE As Integer = vbObjectError + 10
    Private Const DATA_BUFFER As Integer = 10240

    ' Modern classes using BitConverter instead of fixed arrays
    Private Class ByteConverter
        Public ByteArr As Byte()
        Public LongValue As Integer
        Public SingleValue As Single

        Public Sub New()
            ByteArr = New Byte(3) {}
        End Sub
    End Class

    Private Class DoubleConverter
        Public ByteArr As Byte()
        Public DoubleValue As Double

        Public Sub New()
            ByteArr = New Byte(7) {}
        End Sub
    End Class

    Private data() As Byte
    Private queueCapacity As Integer
    Private queueLength As Integer

    Public Sub New()
        data = New Byte(DATA_BUFFER - 1) {}
        queueCapacity = DATA_BUFFER
    End Sub

    Protected Overrides Sub Finalize()
        data = Nothing
        MyBase.Finalize()
    End Sub

    Private Function min(ByVal val1 As Integer, ByVal val2 As Integer) As Integer
        If val1 < val2 Then
            Return val1
        Else
            Return val2
        End If
    End Function

    Private Sub ByteArrayToType(ByRef destVariable As Object, ByRef sourceArray() As Byte, ByVal startPos As Integer, ByVal length As Integer)
        Select Case length
            Case 1
                destVariable = sourceArray(startPos)
            Case 2
                destVariable = BitConverter.ToInt16(sourceArray, startPos)
            Case 4
                If TypeOf destVariable Is Single Then
                    destVariable = BitConverter.ToSingle(sourceArray, startPos)
                Else
                    destVariable = BitConverter.ToInt32(sourceArray, startPos)
                End If
            Case 8
                destVariable = BitConverter.ToDouble(sourceArray, startPos)
        End Select
    End Sub

    Private Sub TypeToByteArray(ByRef destArray() As Byte, ByVal startPos As Integer, ByVal sourceVariable As Object)
        Dim valueType As Type = sourceVariable.GetType()

        If valueType Is GetType(Byte) Then
            destArray(startPos) = CByte(sourceVariable)
        ElseIf valueType Is GetType(Int16) Then
            Dim bytes As Byte() = BitConverter.GetBytes(CShort(sourceVariable))
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length)
        ElseIf valueType Is GetType(Int32) Then
            Dim bytes As Byte() = BitConverter.GetBytes(CInt(sourceVariable))
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length)
        ElseIf valueType Is GetType(Single) Then
            Dim bytes As Byte() = BitConverter.GetBytes(CSng(sourceVariable))
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length)
        ElseIf valueType Is GetType(Double) Then
            Dim bytes As Byte() = BitConverter.GetBytes(CDbl(sourceVariable))
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length)
        End If
    End Sub

    Private Sub CopyArrayData(ByRef destArray() As Byte, ByVal destStart As Integer, ByRef sourceArray() As Byte, ByVal sourceStart As Integer, ByVal length As Integer)
        Array.Copy(sourceArray, sourceStart, destArray, destStart, length)
    End Sub

    Private Function WriteData(ByRef buf() As Byte, ByVal dataLength As Integer) As Integer
        If queueCapacity - queueLength - dataLength < 0 Then
            Throw New InvalidOperationException("Not enough space in the queue")
        End If

        ' Ensure array has enough capacity
        If queueLength + dataLength > data.Length Then
            Array.Resize(Of Byte)(data, data.Length * 2)
            queueCapacity = data.Length
        End If

        CopyArrayData(data, queueLength, buf, 0, dataLength)

        queueLength = queueLength + dataLength
        Return dataLength
    End Function

    Private Function ReadData(ByRef buf() As Byte, ByVal dataLength As Integer) As Integer
        If dataLength > queueLength Then
            Throw New InvalidOperationException("Not enough data in the queue")
        End If

        CopyArrayData(buf, 0, data, 0, dataLength)
        Return dataLength
    End Function

    Private Function ReadDataWithOffset(ByRef buf() As Byte, ByVal dataLength As Integer, ByVal startPos As Integer) As Integer
        If dataLength > queueLength - startPos Then
            Throw New InvalidOperationException("Not enough data in the queue")
        End If

        CopyArrayData(buf, 0, data, startPos, dataLength)
        Return dataLength
    End Function

    Private Function RemoveData(ByVal dataLength As Integer) As Integer
        RemoveData = min(dataLength, queueLength)

        If RemoveData <> queueCapacity Then
            CopyArrayData(data, 0, data, RemoveData, queueLength - RemoveData)
        End If

        queueLength = queueLength - RemoveData
        Return RemoveData
    End Function

    ' Public methods - maintaining the same interface
    Public Function WriteByte(ByVal Value As Byte) As Integer
        Dim buf(0) As Byte

        buf(0) = Value

        Return WriteData(buf, 1)
    End Function

    Public Function WriteInteger(ByVal Value As Short) As Integer
        Dim buf(1) As Byte

        TypeToByteArray(buf, 0, Value)

        Return WriteData(buf, 2)
    End Function

    Public Function WriteLong(ByVal Value As Integer) As Integer
        Dim buf(3) As Byte

        TypeToByteArray(buf, 0, Value)

        Return WriteData(buf, 4)
    End Function

    Public Function WriteSingle(ByVal Value As Single) As Integer
        Dim buf(3) As Byte

        TypeToByteArray(buf, 0, Value)

        Return WriteData(buf, 4)
    End Function

    Public Function WriteDouble(ByVal Value As Double) As Integer
        Dim buf(7) As Byte

        TypeToByteArray(buf, 0, Value)

        Return WriteData(buf, 8)
    End Function

    Public Function WriteBoolean(ByVal Value As Boolean) As Integer
        Dim buf(0) As Byte

        If Value Then buf(0) = 1

        Return WriteData(buf, 1)
    End Function

    Public Function WriteASCIIStringFixed(ByVal Value As String) As Integer
        ' Handle null value
        If Value Is Nothing Then Value = String.Empty

        Dim bytes() As Byte = Text.Encoding.GetEncoding("Windows-1252").GetBytes(Value)
        Return WriteData(bytes, bytes.Length)
    End Function

    Public Function WriteUnicodeStringFixed(ByVal Value As String) As Integer
        ' Handle null value
        If Value Is Nothing Then Value = String.Empty

        Dim bytes() As Byte = Text.Encoding.Unicode.GetBytes(Value)
        Return WriteData(bytes, bytes.Length)
    End Function

    Public Function WriteASCIIString(ByVal Value As String) As Integer
        ' Handle null value
        If Value Is Nothing Then Value = String.Empty

        Dim length As Short = CShort(Value.Length)
        Dim lengthBytes As Byte() = BitConverter.GetBytes(length)
        Dim valueBytes As Byte() = Text.Encoding.GetEncoding("Windows-1252").GetBytes(Value)

        Dim combinedBytes(lengthBytes.Length + valueBytes.Length - 1) As Byte
        Array.Copy(lengthBytes, 0, combinedBytes, 0, lengthBytes.Length)
        Array.Copy(valueBytes, 0, combinedBytes, lengthBytes.Length, valueBytes.Length)

        Return WriteData(combinedBytes, combinedBytes.Length)
    End Function

    Public Function WriteUnicodeString(ByVal Value As String) As Integer
        ' Handle null value
        If Value Is Nothing Then Value = String.Empty

        Dim length As Short = CShort(Value.Length)
        Dim lengthBytes As Byte() = BitConverter.GetBytes(length)
        Dim valueBytes As Byte() = Text.Encoding.Unicode.GetBytes(Value)

        Dim combinedBytes(lengthBytes.Length + valueBytes.Length - 1) As Byte
        Array.Copy(lengthBytes, 0, combinedBytes, 0, lengthBytes.Length)
        Array.Copy(valueBytes, 0, combinedBytes, lengthBytes.Length, valueBytes.Length)

        Return WriteData(combinedBytes, combinedBytes.Length)
    End Function

    Public Function WriteBlock(ByRef Value() As Byte, Optional ByVal length As Integer = -1) As Integer
        ' Handle null array
        If Value Is Nothing Then
            Return 0
        End If

        If length > Value.Length Or length < 0 Then length = Value.Length
        Return WriteData(Value, length)
    End Function

    Public Function ReadByte() As Byte
        Dim buf(0) As Byte

        RemoveData(ReadData(buf, 1))

        Return buf(0)
    End Function

    Public Function ReadInteger() As Short
        Dim buf(1) As Byte
        Dim result As Short

        RemoveData(ReadData(buf, 2))

        result = BitConverter.ToInt16(buf, 0)
        Return result
    End Function

    Public Function ReadLong() As Integer
        Dim buf(3) As Byte
        Dim result As Integer

        RemoveData(ReadData(buf, 4))

        result = BitConverter.ToInt32(buf, 0)
        Return result
    End Function

    Public Function ReadSingle() As Single
        Dim buf(3) As Byte
        Dim result As Single

        RemoveData(ReadData(buf, 4))

        result = BitConverter.ToSingle(buf, 0)
        Return result
    End Function

    Public Function ReadDouble() As Double
        Dim buf(7) As Byte
        Dim result As Double

        RemoveData(ReadData(buf, 8))

        result = BitConverter.ToDouble(buf, 0)
        Return result
    End Function

    Public Function ReadBoolean() As Boolean
        Dim buf(0) As Byte

        RemoveData(ReadData(buf, 1))

        Return buf(0) = 1
    End Function

    Public Function ReadASCIIStringFixed(ByVal length As Integer) As String
        If length <= 0 Then Return String.Empty

        Dim buf(length - 1) As Byte

        If queueLength >= length Then
            RemoveData(ReadData(buf, length))
            Return Text.Encoding.GetEncoding("Windows-1252").GetString(buf)
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function ReadUnicodeStringFixed(ByVal length As Integer) As String
        If length <= 0 Then Return String.Empty

        Dim byteLength As Integer = length * 2
        Dim buf(byteLength - 1) As Byte

        If queueLength >= byteLength Then
            RemoveData(ReadData(buf, byteLength))
            Return Text.Encoding.Unicode.GetString(buf)
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function ReadASCIIString() As String
        If queueLength <= 1 Then
            Throw New InvalidOperationException("Not enough data in the queue")
        End If

        Dim lengthBuf(1) As Byte
        ReadData(lengthBuf, 2)
        Dim length As Short = BitConverter.ToInt16(lengthBuf, 0)

        If queueLength >= length + 2 Then
            RemoveData(2)

            If length > 0 Then
                Dim buf(length - 1) As Byte
                RemoveData(ReadData(buf, length))
                Return Text.Encoding.GetEncoding("Windows-1252").GetString(buf)
            Else
                Return String.Empty
            End If
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function ReadUnicodeString() As String
        If queueLength <= 1 Then
            Throw New InvalidOperationException("Not enough data in the queue")
        End If

        Dim lengthBuf(1) As Byte
        ReadData(lengthBuf, 2)
        Dim length As Short = BitConverter.ToInt16(lengthBuf, 0)
        Dim byteLength As Integer = length * 2

        If queueLength >= byteLength + 2 Then
            RemoveData(2)

            If length > 0 Then
                Dim buf(byteLength - 1) As Byte
                RemoveData(ReadData(buf, byteLength))
                Return Text.Encoding.Unicode.GetString(buf)
            Else
                Return String.Empty
            End If
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function ReadBlock(ByRef block() As Byte, ByVal dataLength As Integer) As Integer
        ' Check if block array is null or not initialized
        If block Is Nothing OrElse block.Length = 0 Then
            Return 0
        End If

        If dataLength > 0 Then
            Return RemoveData(ReadData(block, dataLength))
        Else
            Return 0
        End If
    End Function

    Public Function PeekByte() As Byte
        Dim buf(0) As Byte

        ReadData(buf, 1)

        Return buf(0)
    End Function

    Public Function PeekInteger() As Short
        Dim buf(1) As Byte

        ReadData(buf, 2)

        Return BitConverter.ToInt16(buf, 0)
    End Function

    Public Function PeekLong() As Integer
        Dim buf(3) As Byte

        ReadData(buf, 4)

        Return BitConverter.ToInt32(buf, 0)
    End Function

    Public Function PeekSingle() As Single
        Dim buf(3) As Byte

        ReadData(buf, 4)

        Return BitConverter.ToSingle(buf, 0)
    End Function

    Public Function PeekDouble() As Double
        Dim buf(7) As Byte

        ReadData(buf, 8)

        Return BitConverter.ToDouble(buf, 0)
    End Function

    Public Function PeekBoolean() As Boolean
        Dim buf(0) As Byte

        ReadData(buf, 1)

        Return buf(0) = 1
    End Function

    Public Function PeekASCIIStringFixed(ByVal length As Integer) As String
        If length <= 0 Then Return String.Empty

        Dim buf(length - 1) As Byte

        If queueLength >= length Then
            ReadData(buf, length)
            Return Text.Encoding.GetEncoding("Windows-1252").GetString(buf)
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function PeekUnicodeStringFixed(ByVal length As Integer) As String
        If length <= 0 Then Return String.Empty

        Dim byteLength As Integer = length * 2
        Dim buf(byteLength - 1) As Byte

        If queueLength >= byteLength Then
            ReadData(buf, byteLength)
            Return Text.Encoding.Unicode.GetString(buf)
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function PeekASCIIString() As String
        If queueLength <= 1 Then
            Throw New InvalidOperationException("Not enough data in the queue")
        End If

        Dim lengthBuf(1) As Byte
        ReadData(lengthBuf, 2)
        Dim length As Short = BitConverter.ToInt16(lengthBuf, 0)

        If queueLength >= length + 2 Then
            If length > 0 Then
                Dim buf(length - 1) As Byte
                ReadDataWithOffset(buf, length, 2)
                Return Text.Encoding.GetEncoding("Windows-1252").GetString(buf)
            Else
                Return String.Empty
            End If
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function PeekUnicodeString() As String
        If queueLength <= 1 Then
            Throw New InvalidOperationException("Not enough data in the queue")
        End If

        Dim lengthBuf(1) As Byte
        ReadData(lengthBuf, 2)
        Dim length As Short = BitConverter.ToInt16(lengthBuf, 0)
        Dim byteLength As Integer = length * 2

        If queueLength >= byteLength + 2 Then
            If length > 0 Then
                Dim buf(byteLength - 1) As Byte
                ReadDataWithOffset(buf, byteLength, 2)
                Return Text.Encoding.Unicode.GetString(buf)
            Else
                Return String.Empty
            End If
        Else
            Throw New InvalidOperationException("Not enough data in the queue")
        End If
    End Function

    Public Function PeekBlock(ByRef block() As Byte, ByVal dataLength As Integer) As Integer
        ' Check if block array is null or not initialized
        If block Is Nothing OrElse block.Length = 0 Then
            Return 0
        End If

        If dataLength > 0 Then
            Return ReadData(block, dataLength)
        Else
            Return 0
        End If
    End Function

    Public Sub CopyBuffer(ByRef Source As clsByteQueue)
        ' Check if source is null
        If Source Is Nothing Then
            Exit Sub
        End If

        If Source.length = 0 Then
            RemoveData(length)
            Exit Sub
        End If

        queueCapacity = Source.Capacity

        ReDim data(queueCapacity - 1)

        Dim buf(Source.length - 1) As Byte

        Source.PeekBlock(buf, Source.length)

        queueLength = 0

        WriteBlock(buf, Source.length)
    End Sub

    Public ReadOnly Property length() As Integer
        Get
            Return queueLength
        End Get
    End Property

    Public ReadOnly Property Capacity() As Integer
        Get
            Return queueCapacity
        End Get
    End Property

    Public ReadOnly Property NotEnoughDataErrCode() As Integer
        Get
            Return NOT_ENOUGH_DATA
        End Get
    End Property

    Public ReadOnly Property NotEnoughSpaceErrCode() As Integer
        Get
            Return NOT_ENOUGH_SPACE
        End Get
    End Property
End Class

File: Legacy/Source/clsByteQueue.cs
using System;
using System.Text;

namespace Legacy;

internal class clsByteQueue
{
    // Exception codes
    private const int NOT_ENOUGH_DATA = -2147221495; // vbObjectError + 9
    private const int NOT_ENOUGH_SPACE = -2147221494; // vbObjectError + 10
    private const int DATA_BUFFER = 10240;

    private byte[] data;

    public clsByteQueue()
    {
        data = new byte[10240];
        Capacity = DATA_BUFFER;
    }

    public int length { get; private set; }

    public int Capacity { get; private set; }

    public int NotEnoughDataErrCode => NOT_ENOUGH_DATA;

    public int NotEnoughSpaceErrCode => NOT_ENOUGH_SPACE;

    ~clsByteQueue()
    {
        data = null;
    }

    private int min(int val1, int val2)
    {
        if (val1 < val2) return val1;

        return val2;
    }

    private void ByteArrayToType(ref object destVariable, ref byte[] sourceArray, int startPos, int length)
    {
        switch (length)
        {
            case 1:
            {
                destVariable = sourceArray[startPos];
                break;
            }
            case 2:
            {
                destVariable = BitConverter.ToInt16(sourceArray, startPos);
                break;
            }
            case 4:
            {
                if (destVariable is float)
                    destVariable = BitConverter.ToSingle(sourceArray, startPos);
                else
                    destVariable = BitConverter.ToInt32(sourceArray, startPos);

                break;
            }
            case 8:
            {
                destVariable = BitConverter.ToDouble(sourceArray, startPos);
                break;
            }
        }
    }

    private void TypeToByteArray(ref byte[] destArray, int startPos, object sourceVariable)
    {
        var valueType = sourceVariable.GetType();

        if (ReferenceEquals(valueType, typeof(byte)))
        {
            destArray[startPos] = Convert.ToByte(sourceVariable);
        }
        else if (ReferenceEquals(valueType, typeof(short)))
        {
            var bytes = BitConverter.GetBytes(Convert.ToInt16(sourceVariable));
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length);
        }
        else if (ReferenceEquals(valueType, typeof(int)))
        {
            var bytes = BitConverter.GetBytes(Convert.ToInt32(sourceVariable));
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length);
        }
        else if (ReferenceEquals(valueType, typeof(float)))
        {
            var bytes = BitConverter.GetBytes(Convert.ToSingle(sourceVariable));
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length);
        }
        else if (ReferenceEquals(valueType, typeof(double)))
        {
            var bytes = BitConverter.GetBytes(Convert.ToDouble(sourceVariable));
            Array.Copy(bytes, 0, destArray, startPos, bytes.Length);
        }
    }

    private void CopyArrayData(ref byte[] destArray, int destStart, ref byte[] sourceArray, int sourceStart, int length)
    {
        Array.Copy(sourceArray, sourceStart, destArray, destStart, length);
    }

    private int WriteData(ref byte[] buf, int dataLength)
    {
        if (Capacity - length - dataLength < 0) throw new InvalidOperationException("Not enough space in the queue");

        // Ensure array has enough capacity
        if (length + dataLength > data.Length)
        {
            Array.Resize(ref data, data.Length * 2);
            Capacity = data.Length;
        }

        CopyArrayData(ref data, length, ref buf, 0, dataLength);

        length = length + dataLength;
        return dataLength;
    }

    private int ReadData(ref byte[] buf, int dataLength)
    {
        if (dataLength > length) throw new InvalidOperationException("Not enough data in the queue");

        CopyArrayData(ref buf, 0, ref data, 0, dataLength);
        return dataLength;
    }

    private int ReadDataWithOffset(ref byte[] buf, int dataLength, int startPos)
    {
        if (dataLength > length - startPos) throw new InvalidOperationException("Not enough data in the queue");

        CopyArrayData(ref buf, 0, ref data, startPos, dataLength);
        return dataLength;
    }

    private int RemoveData(int dataLength)
    {
        int RemoveDataRet = default;
        RemoveDataRet = min(dataLength, length);

        if (RemoveDataRet != Capacity) CopyArrayData(ref data, 0, ref data, RemoveDataRet, length - RemoveDataRet);

        length = length - RemoveDataRet;
        return RemoveDataRet;
    }

    // Public methods - maintaining the same interface
    public int WriteByte(byte Value)
    {
        var buf = new byte[1];

        buf[0] = Value;

        return WriteData(ref buf, 1);
    }

    public int WriteInteger(short Value)
    {
        var buf = new byte[2];

        TypeToByteArray(ref buf, 0, Value);

        return WriteData(ref buf, 2);
    }

    public int WriteLong(int Value)
    {
        var buf = new byte[4];

        TypeToByteArray(ref buf, 0, Value);

        return WriteData(ref buf, 4);
    }

    public int WriteSingle(float Value)
    {
        var buf = new byte[4];

        TypeToByteArray(ref buf, 0, Value);

        return WriteData(ref buf, 4);
    }

    public int WriteDouble(double Value)
    {
        var buf = new byte[8];

        TypeToByteArray(ref buf, 0, Value);

        return WriteData(ref buf, 8);
    }

    public int WriteBoolean(bool Value)
    {
        var buf = new byte[1];

        if (Value)
            buf[0] = 1;

        return WriteData(ref buf, 1);
    }

    public int WriteASCIIStringFixed(string Value)
    {
        // Handle null value
        if (Value is null)
            Value = string.Empty;

        var bytes = Encoding.GetEncoding("Windows-1252").GetBytes(Value);
        return WriteData(ref bytes, bytes.Length);
    }

    public int WriteUnicodeStringFixed(string Value)
    {
        // Handle null value
        if (Value is null)
            Value = string.Empty;

        var bytes = Encoding.Unicode.GetBytes(Value);
        return WriteData(ref bytes, bytes.Length);
    }

    public int WriteASCIIString(string Value)
    {
        // Handle null value
        if (Value is null)
            Value = string.Empty;

        var length = Convert.ToInt16(Value.Length);
        var lengthBytes = BitConverter.GetBytes(length);
        var valueBytes = Encoding.GetEncoding("Windows-1252").GetBytes(Value);

        var combinedBytes = new byte[lengthBytes.Length + valueBytes.Length];
        Array.Copy(lengthBytes, 0, combinedBytes, 0, lengthBytes.Length);
        Array.Copy(valueBytes, 0, combinedBytes, lengthBytes.Length, valueBytes.Length);

        return WriteData(ref combinedBytes, combinedBytes.Length);
    }

    public int WriteUnicodeString(string Value)
    {
        // Handle null value
        if (Value is null)
            Value = string.Empty;

        var length = Convert.ToInt16(Value.Length);
        var lengthBytes = BitConverter.GetBytes(length);
        var valueBytes = Encoding.Unicode.GetBytes(Value);

        var combinedBytes = new byte[lengthBytes.Length + valueBytes.Length];
        Array.Copy(lengthBytes, 0, combinedBytes, 0, lengthBytes.Length);
        Array.Copy(valueBytes, 0, combinedBytes, lengthBytes.Length, valueBytes.Length);

        return WriteData(ref combinedBytes, combinedBytes.Length);
    }

    public int WriteBlock(ref byte[] Value, int length = -1)
    {
        // Handle null array
        if (Value is null) return 0;

        if ((length > Value.Length) | (length < 0))
            length = Value.Length;
        return WriteData(ref Value, length);
    }

    public byte ReadByte()
    {
        var buf = new byte[1];

        RemoveData(ReadData(ref buf, 1));

        return buf[0];
    }

    public short ReadInteger()
    {
        var buf = new byte[2];
        short result;

        RemoveData(ReadData(ref buf, 2));

        result = BitConverter.ToInt16(buf, 0);
        return result;
    }

    public int ReadLong()
    {
        var buf = new byte[4];
        int result;

        RemoveData(ReadData(ref buf, 4));

        result = BitConverter.ToInt32(buf, 0);
        return result;
    }

    public float ReadSingle()
    {
        var buf = new byte[4];
        float result;

        RemoveData(ReadData(ref buf, 4));

        result = BitConverter.ToSingle(buf, 0);
        return result;
    }

    public double ReadDouble()
    {
        var buf = new byte[8];
        double result;

        RemoveData(ReadData(ref buf, 8));

        result = BitConverter.ToDouble(buf, 0);
        return result;
    }

    public bool ReadBoolean()
    {
        var buf = new byte[1];

        RemoveData(ReadData(ref buf, 1));

        return buf[0] == 1;
    }

    public string ReadASCIIStringFixed(int length)
    {
        if (length <= 0)
            return string.Empty;

        var buf = new byte[length];

        if (this.length >= length)
        {
            RemoveData(ReadData(ref buf, length));
            return Encoding.GetEncoding("Windows-1252").GetString(buf);
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public string ReadUnicodeStringFixed(int length)
    {
        if (length <= 0)
            return string.Empty;

        var byteLength = length * 2;
        var buf = new byte[byteLength];

        if (this.length >= byteLength)
        {
            RemoveData(ReadData(ref buf, byteLength));
            return Encoding.Unicode.GetString(buf);
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public string ReadASCIIString()
    {
        if (this.length <= 1) throw new InvalidOperationException("Not enough data in the queue");

        var lengthBuf = new byte[2];
        ReadData(ref lengthBuf, 2);
        var length = BitConverter.ToInt16(lengthBuf, 0);

        if (this.length >= length + 2)
        {
            RemoveData(2);

            if (length > 0)
            {
                var buf = new byte[length];
                RemoveData(ReadData(ref buf, length));
                return Encoding.GetEncoding("Windows-1252").GetString(buf);
            }

            return string.Empty;
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public string ReadUnicodeString()
    {
        if (this.length <= 1) throw new InvalidOperationException("Not enough data in the queue");

        var lengthBuf = new byte[2];
        ReadData(ref lengthBuf, 2);
        var length = BitConverter.ToInt16(lengthBuf, 0);
        var byteLength = length * 2;

        if (this.length >= byteLength + 2)
        {
            RemoveData(2);

            if (length > 0)
            {
                var buf = new byte[byteLength];
                RemoveData(ReadData(ref buf, byteLength));
                return Encoding.Unicode.GetString(buf);
            }

            return string.Empty;
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public int ReadBlock(ref byte[] block, int dataLength)
    {
        // Check if block array is null or not initialized
        if (block is null || block.Length == 0) return 0;

        if (dataLength > 0) return RemoveData(ReadData(ref block, dataLength));

        return 0;
    }

    public byte PeekByte()
    {
        var buf = new byte[1];

        ReadData(ref buf, 1);

        return buf[0];
    }

    public short PeekInteger()
    {
        var buf = new byte[2];

        ReadData(ref buf, 2);

        return BitConverter.ToInt16(buf, 0);
    }

    public int PeekLong()
    {
        var buf = new byte[4];

        ReadData(ref buf, 4);

        return BitConverter.ToInt32(buf, 0);
    }

    public float PeekSingle()
    {
        var buf = new byte[4];

        ReadData(ref buf, 4);

        return BitConverter.ToSingle(buf, 0);
    }

    public double PeekDouble()
    {
        var buf = new byte[8];

        ReadData(ref buf, 8);

        return BitConverter.ToDouble(buf, 0);
    }

    public bool PeekBoolean()
    {
        var buf = new byte[1];

        ReadData(ref buf, 1);

        return buf[0] == 1;
    }

    public string PeekASCIIStringFixed(int length)
    {
        if (length <= 0)
            return string.Empty;

        var buf = new byte[length];

        if (this.length >= length)
        {
            ReadData(ref buf, length);
            return Encoding.GetEncoding("Windows-1252").GetString(buf);
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public string PeekUnicodeStringFixed(int length)
    {
        if (length <= 0)
            return string.Empty;

        var byteLength = length * 2;
        var buf = new byte[byteLength];

        if (this.length >= byteLength)
        {
            ReadData(ref buf, byteLength);
            return Encoding.Unicode.GetString(buf);
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public string PeekASCIIString()
    {
        if (this.length <= 1) throw new InvalidOperationException("Not enough data in the queue");

        var lengthBuf = new byte[2];
        ReadData(ref lengthBuf, 2);
        var length = BitConverter.ToInt16(lengthBuf, 0);

        if (this.length >= length + 2)
        {
            if (length > 0)
            {
                var buf = new byte[length];
                ReadDataWithOffset(ref buf, length, 2);
                return Encoding.GetEncoding("Windows-1252").GetString(buf);
            }

            return string.Empty;
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public string PeekUnicodeString()
    {
        if (this.length <= 1) throw new InvalidOperationException("Not enough data in the queue");

        var lengthBuf = new byte[2];
        ReadData(ref lengthBuf, 2);
        var length = BitConverter.ToInt16(lengthBuf, 0);
        var byteLength = length * 2;

        if (this.length >= byteLength + 2)
        {
            if (length > 0)
            {
                var buf = new byte[byteLength];
                ReadDataWithOffset(ref buf, byteLength, 2);
                return Encoding.Unicode.GetString(buf);
            }

            return string.Empty;
        }

        throw new InvalidOperationException("Not enough data in the queue");
    }

    public int PeekBlock(ref byte[] block, int dataLength)
    {
        // Check if block array is null or not initialized
        if (block is null || block.Length == 0) return 0;

        if (dataLength > 0) return ReadData(ref block, dataLength);

        return 0;
    }

    public void CopyBuffer(ref clsByteQueue Source)
    {
        // Check if source is null
        if (Source is null) return;

        if (Source.length == 0)
        {
            RemoveData(length);
            return;
        }

        Capacity = Source.Capacity;

        data = new byte[Capacity];

        var buf = new byte[Source.length];

        Source.PeekBlock(ref buf, Source.length);

        length = 0;

        WriteBlock(ref buf, Source.length);
    }

    // Modern classes using BitConverter instead of fixed arrays
    private class ByteConverter
    {
        public byte[] ByteArr;

        public ByteConverter()
        {
            ByteArr = new byte[4];
        }
    }

    private class DoubleConverter
    {
        public byte[] ByteArr;

        public DoubleConverter()
        {
            ByteArr = new byte[8];
        }
    }
}